### ISSUE TITLE
Switch statement with always matching declaration pattern case and 'default' case: unreachable code is not reported

### ISSUE THREAD

**Issue by controlflow** [dotnet/roslyn#15138]
```c#
  public bool F(int i) {
    switch (i) {
      case int t:
        return true;
      default:
        return false; // unreachable, but not reported
    }
  
    return false; // unreachable
  }
```

It is safe to introduce compiler warning there, since this missing unreachable code case can only happen because of declaration patterns (available only in C# 7.0). One more example:

```c#
  public bool F(string s) {
    switch (s) {
      case null:
      case object o:
        return true;
      default:
        return false; // unreachable, but not reported
    }

    return false; // unreachable
  }
```

**Comment by gafter** [comment]
The unreachable case could happen before, for example if you were switching on a bool and had cases for true and false, and we never reported a warning for that before. We intentionally treat the default case as always reachable.
